tso-server/server/server.go

```diff
@@ -131,60 +204,240 @@ type session struct {
 	conn net.Conn
 }
 
-// NewTimestampOracle creates a tso server with listen address addr.
-func NewTimestampOracle(addr string) (*TimestampOracle, error) {
-	tso := &TimestampOracle{
-		ticker: time.NewTicker(time.Duration(updateTimtestampStep) * time.Millisecond),
+// NewTimestampOracle creates a tso server with special config.
+func NewTimestampOracle(cfg *Config) (*TimestampOracle, error) {
+	if cfg.SaveInterval <= 0 {
+		cfg.SaveInterval = defaultSaveInterval
 	}
-	current := &atomicObject{
-		physical: time.Now(),
+
+	tso := &TimestampOracle{
+		cfg:      cfg,
+		isLeader: 0,
 	}
-	tso.ts.Store(current)
-	go tso.updateTicker()
 
-	ln, err := net.Listen("tcp", addr)
+	var err error
+	tso.listener, err = net.Listen("tcp", cfg.Addr)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	tso.listener = ln
+	if len(cfg.ZKAddr) == 0 {
+		// no zookeeper, use a fake zk conn instead
+		tso.zkConn = zkhelper.NewConn()
+	} else {
+		tso.zkConn, err = zkhelper.ConnectToZkWithTimeout(cfg.ZKAddr, zkTimeout)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
 	tso.conns = map[net.Conn]struct{}{}
 
 	return tso, nil
 }
 
-// Run runs the timestamp oracle server.
-func (tso *TimestampOracle) Run() {
+type tsoTask struct {
+	tso *TimestampOracle
+
+	connCh chan net.Conn
+
+	interruptCh chan error
+	interrupted int64
+
+	stopCh chan struct{}
+}
+
+func (t *tsoTask) Run() error {
+	if t.Interrupted() {
+		return errors.New("task has been interrupted already")
+	}
+
+	// enter here means I am the leader tso.
+	tso := t.tso
+
+	log.Debugf("tso leader %s run task", tso)
+
+	if err := tso.loadTimestamp(); err != nil {
+		// should we interrupt the task here now?
+		return errors.Trace(err)
+	}
+
+	atomic.StoreInt64(&tso.isLeader, 1)
+
+	tsTicker := time.NewTicker(time.Duration(updateTimtestampStep) * time.Millisecond)
+
+	defer func() {
+		atomic.StoreInt64(&tso.isLeader, 0)
+
+		tsTicker.Stop()
+
+		// we should close all connections.
+		t.closeAllConns()
+		log.Debugf("tso leader %s task end", tso)
+	}()
+
+	for {
+		var conn net.Conn
+
+		select {
+		case err := <-t.interruptCh:
+			log.Debugf("tso leader %s is interrupted, err %v", tso, err)
+
+			// we will interrupt this task, and we can't run this task again.
+			atomic.StoreInt64(&t.interrupted, 1)
+			return errors.Trace(err)
+		case <-t.stopCh:
+			log.Debugf("tso leader %s is stopped", tso)
+
+			// we will stop this task, maybe run again later.
+			return errors.New("task is stopped")
+		case conn = <-t.connCh:
+			// handle connection below
+
+			s := &session{
+				r:    bufio.NewReaderSize(conn, sessionReadBufferSize),
+				w:    bufio.NewWriterSize(conn, sessionWriteBufferSize),
+				conn: conn,
+			}
+
+			tso.addConn(conn)
+			tso.wg.Add(1)
+			go tso.handleConnection(s)
```

Comment:
Oh I figured it out. All connections will be closed by `tsoTask` before it ends.